design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+`spec.package` field or by changing another field such as `spec.source`, should
+result in a new `PackageRevision` unless one with the same name already exists.
+`PackageRevision`s that are produced by the _`Package` controller_ will also be
+labeled with their parent `Package`'s name.
+
+_Note: in a backup and restore scenario, a `Package` should be able to "regain
+control" of its `PackageRevision`s by querying based on labels. If a
+`PackageRevision` with the appropriate name does not exist, the _`Package`
+controller_ knows that it needs to create a new one._
+
+The workflow for the _`Package` controller_ will be as follows:
+
+*New `Package` Created*
+
+1. Observe creation of a `Package`.
+2. Create install `Job` with the same name as the `Package` that produces
+   manifests for a `PackageRevision` and associated objects.
+3. Create the `PackageRevision` with a name equivalent to the `sha256` hash of
+   the package image, is owned by the `Package`, has a `spec.state` of `Active`,
+   and is labeled with the name of the `Package`.  The `sha256` hash of the
+   package image can be observed from the `initContainerStatuses[0].imageID` of
+   the install `Job`.
+   - If a `PackageRevision` exists with name equivalent to `sha256` hash of the
+     package image, but it is not owned by this `Package`, we are unable to
+     install.
+4. Wait for `PackageRevision` to become `Ready` (this will occur after
+   successful install of all objects included in the `PackageRevision`, see the
+   _`PackageRevision` controller_ section for more details).
+5. Set the `status.currentHash` of the `Package` to match the `sha256` hash of
+   the package image and `status.currentImage` to match the package image name.
+
+*Existing `Package` Updated*
+
+1. Observe update of a `Package`.
+  - If the `spec.package` image does not include an appended hash and matches
+    the `status.currentImage`, no change is needed assuming the current
+    `PackageRevision` is healthy.
+  - If the user supplies an image with hash and both the `status.currentImage`
+    and `status.currentHash` match the `spec.package`, no change is needed,
+    assuming the current `PackageRevision` is healthy.
+  - If the user supplies an image with hash and either  the
+    `status.currentImage` and `status.currentHash` do not match the
+    `spec.package`, check to see if an existing install `Job` with same name as
+    the `Package` exists and has the same image and hash in the `initContainer`.
+    If so, no change is needed, assuming the current `PackageRevision` is
+    healthy.
+  - If the user supplies an image with hash and either the the
+    `status.currentImage` and `status.currentHash` do not match the
+    `spec.package` and does not match an existing install `Job` (or existing
+    install `Job` does not exist), the controller will proceed with update.
+2. Delete any existing install `Job` with the same name as the `Package` and
+   create a new one with image in `spec.package`.
```

Comment:
was just thinking about the lifecycle of the install/unpacking job also, its always bothered me that we just leave them hanging around.  Is there a reason to do that at all? Can we delete the job every time as soon as we read the unpacked output from its pod logs?

Reply:
@negz this is not possible in the this proposal because one of the purposes of invoking the install `Job` is to get the image hash, which is then used to name the `PackageRevision`. Also in this proposal we don't really have to garbage collect jobs. There is only ever one existing `Job` at a time for a `Package`, and it has the same name. If the `Package` has an owner reference on this `Job` then it will be cleaned up when the `Package` is deleted.

@jbw976 we could go ahead and clean them up after the corresponding `PackageRevision` is set to `Active` and is `Ready` 👍

Reply:
> There is only ever one existing Job at a time for a Package, and it has the same name. If the Package has an owner reference on this Job then it will be cleaned up when the Package is deleted.

Does that mean at least one `Job` always has to be there? In general, I think we shouldn't depend on existence of a `Completed` `Job` because if a `Job` is `Completed`, IMO it implicitly declares that it has no purpose in the cluster hence safe to delete, but we'd be giving it one.

Reply:
@muvaf We could clean them up immediately

Reply:
> We could clean them up immediately

@hasheddan Do you mean we can clean it up after the installation or during the removal? Does it not have to exist while the package installed?

Reply:
After installation. Once a `PackageRevision` is `Active` and `Ready` it should have all information needed in its spec, except for the contents of the objects it brought, which will have to have been successfully created if the `PackageRevision` is ready.